packages/yarnpkg-core/sources/Manifest.ts

```diff
@@ -350,12 +350,17 @@ export class Manifest {
         ? data.workspaces.packages
         : [];
 
+    const existingWorkspaces = new Set(this.workspaceDefinitions.map(({pattern}) => pattern));
```

Comment:
Couldn't we just reset `this.workspaceDefinitions` to an empty array? This way it wouldn't change its type / content.

Reply:
That's also possible.

There are some other problems with the `Manifest#load` function, e.g. it resets some properties but it doesn't reset `dependencies`/`optionalDependencies`/`peerDependencies`/`devDependencies`/`dependenciesMeta`/`peerDependenciesMeta` and `resolutions`. This means it's possible to add/update values in these objects, but impossible to remove them by reloading raw JSON data. The load method also doesn't support removing `preferUnplugged`.

All of those properties should be reset, otherwise we'll just end up getting more bugs like "constraints can't remove resolutions/dependenciesMeta/peerDependenciesMeta/preferUnplugged... via `gen_enforced_field`"

In short, this method wasn't written to be used on an already loaded manifest, yielding all kind of shenanigans in constraints because we do use it on an already loaded manifest.